### PR TITLE
[luci/pass] Add shape inference to RemoveUnneccessaryTransposeNetPass

### PR DESCRIPTION
**Pull request by zetwhite**
This commit extends RemoveUnncessaryTransposeNetPass to handle reshape/shape with -1 value.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

related : https://github.com/Samsung/ONE/issues/12266

### CODE REVIEW

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 1]
```diff
@@ -82,16 +82,61 @@ class TaggedShapeAnalyzer final
...
+  for (const auto &dim : shape)
+  {
+    if (dim.value >= 1)
+      size *= dim.value;
+  }
+  return size;
+}
+
+/**
+ * @brief if 'dst' has -1 valued dim, replace -1 with infereced value
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:110]:
```suggestion
 * @brief if 'dst' has -1 valued dim, replace -1 with inferenced value
```

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 2]
```diff
@@ -82,16 +82,61 @@ class TaggedShapeAnalyzer final
...
+
+  if (incomplete_indexes.size() == 0)
+    return true;
+  else if (incomplete_indexes.size() == 1)
+  {
+    auto &dim = dst[incomplete_indexes[0]];
+    if (flatsize(src) % flatsize(dst) == 0)
+      dim.value = flatsize(src) / flatsize(dst);
+    else
+      return false;
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:132]:
```suggestion
    if (flatsize(src) % flatsize(dst) == 0)
      dst[incomplete_indexes[1]].value = flatsize(src) / flatsize(dst);
    else
      return false;
```

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 3]
```diff
@@ -176,14 +221,17 @@ bool TaggedShapeAnalyzer::analyze_reshape(const luci::CircleReshape *reshape_nod
   {
     TaggedShapeAnalyzer::Dim dim;
     dim.value = shape_node->at<ReshapeType>(i);
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:223]:
```suggestion
    dim.value = shape_node->at<ReshapeType>(i);
    
```

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 4]
```diff
@@ -176,14 +221,17 @@ bool TaggedShapeAnalyzer::analyze_reshape(const luci::CircleReshape *reshape_nod
...
-
     new_shape.push_back(dim);
   }
 
+  // inferece new_shape's -1 dim
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:227]:
```suggestion
  // inference new_shape dim with -1 value
```

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 5]
```diff
@@ -82,16 +82,61 @@ class TaggedShapeAnalyzer final
 
   struct Dim final
   {
-    uint32_t value;
+    int32_t value;
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:85]: (NOTE) reshape/shape value can have -1 value.

ref / pytorch reshape said 'A single dimension can be -1' [:link:]( https://pytorch.org/docs/master/generated/torch.reshape.html#torch.reshape:~:text=A%20single%20dimension%20may%20be%20%2D1%2C%20in%20which%20case%20it%E2%80%99s%20inferred%20from%20the%20remaining%20dimensions%20and%20the%20number%20of%20elements%20in%20input.)

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp [hunk 6]
```diff
@@ -32,7 +32,7 @@ class TransposeReshapeTransposeGraph : public TestIOGraph
 
 public:
   // create input-transpose-reshape-transpose-output graph
-  void init_whole_graph(ShapeU32 in_shape, ShapeU32 front_perm, ShapeU32 mid_shape,
+  void init_whole_graph(ShapeU32 in_shape, ShapeU32 front_perm, ShapeI32 mid_shape,
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp:35]: (NOTE) reshape/shape can have -1 value.

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp [hunk 7]
```diff
@@ -82,16 +82,60 @@ class TaggedShapeAnalyzer final
...
+    if (dst[i].value == -1)
+      incomplete_indexes.push_back(i);
+  }
+
+  if (incomplete_indexes.size() == 0)
+    return true;
+  else if (incomplete_indexes.size() == 1)
+  {
+    if (flatsize(src) % flatsize(dst) == 0)
+      dst[incomplete_indexes[1]].value = flatsize(src) / flatsize(dst);
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp:129]:
```suggestion
      dst[incomplete_indexes[0]].value = flatsize(src) / flatsize(dst);
```

File: compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp [hunk 8]
```diff
@@ -217,6 +228,27 @@ TEST(RemoveUnnecessaryTransposeNetPass, identity_pattern)
   EXPECT_TRUE(is_transpose_removed(g.g()));
 }
 
+TEST(RemoveUnnecessaryTransposeNetPass, incomplete_reshape_pattern)
```
Comment (zetwhite) [inline compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp:231]: Newly added 3 tests passed.
I also checked our target model(efficient former) is optimized well as we expected.
```sh 
➜ ./build/compiler/luci/pass/luci_pass_test --gtest_filter='RemoveUnnecessaryTransposeNetPass.incomplete*'
Running main() from /home/sseung/Workspace/ONE_second/externals/GTEST/googletest/src/gtest_main.cc
Note: Google Test filter = RemoveUnnecessaryTransposeNetPass.incomplete*
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from RemoveUnnecessaryTransposeNetPass
[ RUN      ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern
[       OK ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern (0 ms)
[ RUN      ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern1_NEG
[       OK ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern1_NEG (0 ms)
[ RUN      ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern2_NEG
[       OK ] RemoveUnnecessaryTransposeNetPass.incomplete_reshape_pattern2_NEG (0 ms)
[----------] 3 tests from RemoveUnnecessaryTransposeNetPass (0 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (0 ms total)
[  PASSED  ] 3 tests.
```

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]
LGTM

**Pull request merged**